### PR TITLE
DMP-483 initial implementation

### PR DESCRIPTION
**Pull request by garethmorriscgi**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DMP-483 

### Change description ###
Link audio that has not already been linked to a hearing that is associated with hearings for event/case logs that occurred within a 30 minute window of the audio start and end time.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/darts/testutils/stubs/EventStub.java [hunk 1]
```diff
@@ -38,4 +38,19 @@ public EventEntity createEvent(HearingEntity hearing, int eventHandlerId) {
         return eventEntity;
     }
 
+    public EventEntity createEvent(HearingEntity hearing, int eventHandlerId, OffsetDateTime eventTimestamp) {
```
Comment (johnrace0) [inline src/integrationTest/java/uk/gov/hmcts/darts/testutils/stubs/EventStub.java:41]: maybe could combine this with existing createEvent method?

File: src/test/java/uk/gov/hmcts/darts/audio/service/impl/AudioServiceImplTest.java [hunk 2]
```diff
@@ -218,6 +230,48 @@ void addAudio() throws IOException {
...
+        when(audioConfigurationProperties.getHandheldAudioCourtroomNumbers())
+            .thenReturn(Arrays.asList(addAudioMetadataRequest.getCourtroom()));
+
+        audioService.linkAudioToHearingByEvent(addAudioMetadataRequest, mediaEntity);
+        verify(hearingRepository, times(0)).saveAndFlush(any());
+        assertEquals(0, hearing.getMediaList().size());
+    }
+
+    @Test
+    void linkAudioToHearingByEvent() {
```
Comment (johnrace0) [inline src/test/java/uk/gov/hmcts/darts/audio/service/impl/AudioServiceImplTest.java:254]: does this test cover pre-amble only? is it worth having a test for post-amble / both?

**Review by johnrace0 — CHANGES_REQUESTED** [review]

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -116,6 +116,10 @@ darts:
       - "mpeg2"
       - "mp2"
     max-file-size: 524288000
+    preAmbleSize: 30
+    postAmbleSize: 30
```
[the remark this answers is not in the document]
Comment (garethmorriscgi) [reply src/main/resources/application.yaml]: Good suggestion - have updated the names

File: src/test/java/uk/gov/hmcts/darts/audio/service/impl/AudioServiceImplTest.java (hunk 2 above)
Comment (garethmorriscgi) [reply src/test/java/uk/gov/hmcts/darts/audio/service/impl/AudioServiceImplTest.java:254]: The response from searching for events is mocked so a post-amble test would in theory be a duplicate. I have however added an additional test to cover when the query brings back more than one court log for the same hearing to check that the audio is only linked once - and have used post amble times for the court log events.

File: src/integrationTest/java/uk/gov/hmcts/darts/testutils/stubs/EventStub.java (hunk 1 above)
Comment (garethmorriscgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/testutils/stubs/EventStub.java]: I've amended the method so the others call the newly created one

File: src/main/java/uk/gov/hmcts/darts/audio/config/AudioConfigurationProperties.java [hunk 4]
```diff
@@ -34,5 +34,7 @@ public class AudioConfigurationProperties {
     private List<String> allowedMediaFormats = new ArrayList<>();
     private Integer maxFileSize;
     private Duration allowableAudioGapDuration;
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/audio/config/AudioConfigurationProperties.java:36]: @garethmorriscgi This is an example of a `Duration` field in the config, like what @johnrace0 was mentioning earlier

**Review by johnrace0 — CHANGES_REQUESTED** [review]
re: comment from RB about switching from Integer to Duration. is this something that can/will be done for this ticket?

**Comment by garethmorriscgi** [comment]
> re: comment from RB about switching from Integer to Duration. is this something that can/will be done for this tic

The way that this configuration value is being used i.e. reducing/adding minutes to an OffsetDateTime which expects a long) I don't see the need to use the Duration type here. You'd need to do unnecessary additional work and lose code readability.

**Review by johnrace0 — APPROVED** [review]

**Pull request merged**